File: packages/shared-py/olma_shared/bus.py
"""Event bus abstraction (instructions.md §8.2, Principle #6/#8).

Prod is NATS JetStream (replay, at-least-once); this defines the interface plus
an in-process implementation for dev/tests. Delivery is at-least-once, so a
DedupeGuard lets consumers drop repeats by message_id / idempotency_key (§21).
Subjects support a single trailing '*' wildcard (e.g. 'agent.events.*').
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
from typing import Any, Awaitable, Callable, Protocol


@dataclass
class Message:
    subject: str
    data: dict[str, Any]
    message_id: str = ""  # dedup key (message_id or idempotency_key)


Handler = Callable[[Message], Awaitable[None]]


class Bus(Protocol):
    async def publish(self, subject: str, data: dict, message_id: str = "") -> None: ...
    def subscribe(self, pattern: str, handler: Handler) -> Callable[[], None]: ...


def _matches(pattern: str, subject: str) -> bool:
    if pattern == subject:
        return True
    if pattern.endswith(".*"):
        return subject.startswith(pattern[:-1])  # 'a.b.*' matches 'a.b.<x>'
    return False


@dataclass
class InMemoryBus:
    """In-process fan-out bus. One instance = one 'cluster' (dev only)."""

    _subs: list[tuple[str, Handler]] = field(default_factory=list)

    async def publish(self, subject: str, data: dict, message_id: str = "") -> None:
        msg = Message(subject=subject, data=data, message_id=message_id)
        handlers = [h for pat, h in self._subs if _matches(pat, subject)]
        # at-least-once: deliver to every matching subscriber
        await asyncio.gather(*(h(msg) for h in handlers)) if handlers else None

    def subscribe(self, pattern: str, handler: Handler) -> Callable[[], None]:
        entry = (pattern, handler)
        self._subs.append(entry)

        def unsubscribe() -> None:
            if entry in self._subs:
                self._subs.remove(entry)

        return unsubscribe


class DedupeGuard:
    """Consumer-side at-least-once dedup by message_id (§21)."""

    def __init__(self, capacity: int = 10000) -> None:
        self._seen: dict[str, None] = {}
        self._capacity = capacity

    def is_duplicate(self, message_id: str) -> bool:
        if not message_id:
            return False
        if message_id in self._seen:
            return True
        self._seen[message_id] = None
        if len(self._seen) > self._capacity:
            # drop oldest (dict preserves insertion order)
            oldest = next(iter(self._seen))
            del self._seen[oldest]
        return False

File: packages/shared-ts/src/bus.ts
// Event bus abstraction (instructions.md §8.2, Principle #6/#8).
// Wire-compatible contract with packages/shared-py/olma_shared/bus.py: subjects
// use a single trailing '*' wildcard; delivery is at-least-once, so a DedupeGuard
// lets consumers drop repeats by message_id. Prod is NATS JetStream.

export interface Message {
  subject: string;
  data: Record<string, unknown>;
  messageId: string;
}

export type Handler = (msg: Message) => Promise<void> | void;

function matches(pattern: string, subject: string): boolean {
  if (pattern === subject) return true;
  if (pattern.endsWith(".*")) return subject.startsWith(pattern.slice(0, -1));
  return false;
}

export class InMemoryBus {
  private subs: Array<[string, Handler]> = [];

  async publish(subject: string, data: Record<string, unknown>, messageId = ""): Promise<void> {
    const msg: Message = { subject, data, messageId };
    const handlers = this.subs.filter(([pat]) => matches(pat, subject)).map(([, h]) => h);
    await Promise.all(handlers.map((h) => h(msg)));
  }

  subscribe(pattern: string, handler: Handler): () => void {
    const entry: [string, Handler] = [pattern, handler];
    this.subs.push(entry);
    return () => {
      const i = this.subs.indexOf(entry);
      if (i >= 0) this.subs.splice(i, 1);
    };
  }
}

// Consumer-side at-least-once dedup by message_id (§21).
export class DedupeGuard {
  private seen = new Set<string>();
  private capacity: number;
  constructor(capacity = 10000) {
    this.capacity = capacity;
  }

  isDuplicate(messageId: string): boolean {
    if (!messageId) return false;
    if (this.seen.has(messageId)) return true;
    this.seen.add(messageId);
    if (this.seen.size > this.capacity) this.seen.delete(this.seen.values().next().value!);
    return false;
  }
}
